components/batch-system/src/router.rs

```diff
@@ -272,14 +300,49 @@ where
         info!("[region {}] shutdown mailbox", addr);
         unsafe { &mut *self.caches.as_ptr() }.remove(&addr);
         let mut mailboxes = self.normals.lock().unwrap();
-        if let Some(mb) = mailboxes.remove(&addr) {
+        if let Some(mb) = mailboxes.map.remove(&addr) {
             mb.close();
         }
+        mailboxes
+            .alive_cnt
+            .store(mailboxes.map.len(), Ordering::Relaxed);
     }
 
     pub fn clear_cache(&self) {
         unsafe { &mut *self.caches.as_ptr() }.clear();
     }
+
+    pub fn state_cnt(&self) -> &Arc<AtomicUsize> {
+        &self.state_cnt
+    }
+
+    pub fn alive_cnt(&self) -> Arc<AtomicUsize> {
+        self.normals.lock().unwrap().alive_cnt.clone()
+    }
+
+    pub fn trace(&self) -> RouterTrace {
+        let alive = self.normals.lock().unwrap().alive_cnt.clone();
+        let total = self.state_cnt.load(Ordering::Relaxed);
+        let alive = alive.load(Ordering::Relaxed);
+        let leak = if total > alive + 1 {
+            total - alive - 1
+        } else {
+            0
+        };
+        let mailbox_unit = mem::size_of::<(u64, BasicMailbox<N>)>();
+        let state_unit = mem::size_of::<FsmState<N>>();
+        // Every message in crossbeam sender needs 8 bytes to store state.
+        let message_unit = mem::size_of::<N::Message>() + 8;
+        // crossbeam unbounded channel sender has a list of blocks. Every block has 31 unit
+        // and every sender has at least one sender.
+        let sender_block_unit = 31;
+        RouterTrace {
+            alive: alive * mailbox_unit * 8 / 7 // hashmap uses 7/8 of allocated memory.
```

Comment:
8/7 or 7/8?

Reply:
It uses 7/8 memory so we need to multiple 8/7 to get actually allocated size.

Reply:
got it. BTW, what not `(mailbox_unit * 8 / 7 + state_unit + message_unit * sender_block_unit) * alive`